### PR TITLE
Feature: manage package

### PR DESCRIPTION
**Pull request by jgnagy**
Allowing a user to decline the installation of the owncloud packages. This is pretty typical across puppet modules. Choosing to not install the package via `manage_package = false` also sets `manage_repo` to `false` for simplicity.

### CODE REVIEW

**Comment by shoekstra** [comment]
How would the package be installed if not from the module?

**Comment by jgnagy** [comment]
If you're using a different repository that uses different package names, or if you used the tarball installation method with updates from within the app rather than from a repo, or if you used the "web install", or if you want to target / lock to a specific version, or possibly enterprise (though I'm not sure if that can be installed manually). I'm sure there are more reasons, and I don't think this level of flexibility comes at much cost or maintenance effort, but I obviously leave that to you to decide.

**Comment by shoekstra** [comment]
While I can see the benefit of exposing the package version (and maybe name) I cannot see how choosing not to install the package would benefit; additionally installing and upgrading using packages is the recommended way to manage ownCloud installations as per their documentation.

Re other methods, the way they've redone the ownCloud packages in version 9 may solve this.  They now have two packages, `owncloud-deps` and `owncloud-files`; the former installs dependencies and the latter just the ownCloud files.  I think providing a way to install or not install the deps package via this module and still installing the `owncloud-files` package is the way to go.  This gives you the freedom to install and manage dependencies yourself if you choose but still provides recommended upgrade path via package using `owncloud-files`.

**Comment by jgnagy** [comment]
Given that I wanted to run version 9.x and this module didn't offer a way to do that, this feature was by far the easiest way to enable that. You can handle the choosing of the package version (and the complexities associated with changing package names from the vendor) however you'd like, but until that happens, it doesn't seem to make sense to restrict your module users to older versions. This is just one such example; there may be many cases where a user can benefit from this module setting up ownCloud (config, apache, etc.) without needing it to install the OS package. After all, who knows what will happen with version 10?

Even if the above case wasn't valid, and while I agree that using packages when available is ownCloud's preferred installation method, package-based installation clearly isn't their _only_ supported method. They have extensive documentation on how to install from source. I'm not sure why there is push-back to allowing users to use whichever supported (or even unsupported) installation they choose. Given that this isn't a request for you to do more work, and instead is a PR with the work already done, I find this even more confusing. Ultimately, offering the user of your module more options while providing sane / intuitive defaults is what makes software flexible. I currently use this option in production, so understandably from my perspective, you not seeing the utility to a `manage_package` option doesn't make a compelling case against it. I get that, in theory, all practical use-cases for different package versions and installation sources could be accounted for explicitly with enough effort, thereby making this option less appealing, in practice this is usually slow and the wait is usually frustrating.

Finally, one other thing to consider: offering the ability to install from source, along with some simple additions such as being able to pass in the `$datadirectory` and `$documentroot`, and maybe using a defined type rather than a class would be enough to allow users to run multiple versions of ownCloud on the same machine (say for testing new features, multi-tenancy, etc.).

In summary, I understand that it is your prerogative to decide which features your module will (and will not) support, but please don't limit what is arguably a fairly standard puppet parameter simply on the basis that you can't see its utility given your use case, especially when the request is coming from someone who clearly can.

**Comment by shoekstra** [comment]
The idea behind having the `$manage_*` parameters was to allow the user to use services in place of the ones supported by this module; for example one would set `$manage_apache` to `false` so that they could install NGINX in it's place (or perhaps swap out MySQL for Postgres) and ideally port that functionality back to this module. What you've done is provide an option for the ownCloud module to not install ownCloud - I can't get my head around that. How would the package get installed?  The user installs it by hand?  That kinda goes against what a system under config management is all about IMHO. If a Puppet module cannot support your way of managing the app or service it makes more sense to fix the module than to provide an option to disable it and configure the server by hand.

Re installing from source or package: it shouldn't matter what method is used if the end result is the same - why give multiple options to do the same thing? Since there are many ways to do the same thing, it seems logical to use the recommended way unless the other methods offer more options (or perhaps you install from an SCM like a git or svn, then it's useful to have that option so you can get updates/patches quicker without waiting for a package to be released).

Re multiple instances, this has been considered but at the time of writing this initially it was made for a single instance (there is actually a note at the bottom of the README that suggests someone could convert the class to a definition to support multi-tenant environments). There is no requirement to install from source to have multiple instances - per instance you need a website config (for Apache or NGINX), an ownCloud config file, a separate database or tablespace. I'm going through the module now to make it functional for version 9 (let's hope they don't decide to change how they package it again in version 10!) so I could change to definitions to support multiple tenants.

Re allowing users to run multiple versions on the same machine, this is possibly the worst idea - if you want to test different versions you should do so on in a test or acceptance environment, not on a production/live system.

I really do try hard to accommodate all community suggestions and requests but I am reluctant to accept this PR and would rather refactor the module to future proof it (e.g. allow providing package names incase the maintainers change their mind again) than encourage installing ownCloud by another means than this Puppet module.
